### ISSUE TITLE
Can I combine toPorts and fromCIDR to whitelist external IP addresses for specified ports?

### ISSUE THREAD

**Issue by moredure** [cilium/cilium#8426]
I am using DigitalOcean's CiliumNetworkPolicy, and trapped into not able to use cidr with both toPorts or ports (using default kubernetes network policy resource definition) 
```
---
kind: CiliumNetworkPolicy
apiVersion: "cilium.io/v2"
metadata:
  name: web-allow-external1
spec:
  endpointSelector:
    matchLabels:
      app: nginx
  ingress:
    - fromCIDR:
        - "85.238.106.86/32"
      fromPorts:
        - ports:
            - port: "6379"
              protocol: TCP
  egress:
    - {}
```

**Comment by mikefaraponov** [comment]
Also I expect that in such cases
```
kind: CiliumNetworkPolicy
apiVersion: "cilium.io/v2"
metadata:
  name: web-allow-external2
spec:
  endpointSelector:
    matchLabels:
      app: nginx
  ingress:
    - fromCIDR:
          - "85.238.106.89/32"
      fromPorts:
      - ports:
          - port: "6379"
            protocol: TCP
    - fromCIDR:
        - "62.16.0.182/32"
      fromPorts:
        - ports:
            - port: "80"
              protocol: TCP
  egress:
    - {}
```
I shouldn't be able to access 6379 from both specified ips

**Comment by aanm** [comment]
Hello @mikefaraponov, is your expectation that port `6379` should be allowed from `"62.16.0.182/32"` **and** `"85.238.106.89/32"`?

**Comment by mikefaraponov** [comment]
No, I expect that 6379 should only be allowed from 85.238.106.89/32

**Comment by aanm** [comment]
@mikefaraponov ah thanks! yeah the issue seems to be a duplicate of this one https://github.com/cilium/cilium/issues/8231 This has been fixed and will be released in v1.5.4 and v1.4.5.

**Comment by gandro** [comment]
@aanm I'm not sure this is an instance of #8231, as that issue was related to the translation of vanilla K8s `NetworkPolicy` to `CiliumNetworkPolicy` rules. Here we are submitting `CiliumNetworkPolicy` rules directly, so I believe this is a different cause.

**Comment by mikefaraponov** [comment]
I also experience same issue with plain networkpolicy (and I am not sure which version of cillium DO is using in their 1.14 k8s version )

**Comment by gandro** [comment]
So yes, it seems like policies with FromCIDR and ToPorts are currently not supported by cilium yet.

Applying the yaml from https://github.com/cilium/cilium/issues/8426#issuecomment-506968414 yields a `"Combining FromCIDR and ToPorts is not supported yet"` error in the log output in cilium-agent:

```
Jul 02 13:18:58 k8s1 cilium-agent[18187]: level=warning msg="Error parsing new CiliumNetworkPolicy rule" error="Invalid CiliumNetworkPolicy spec: Combining FromCIDR and ToPorts is not supported yet" obj="TypeMeta: &TypeMeta{Kind:CiliumNetworkPolicy,APIVersion:cilium.io/v2,}, ObjectMeta: &ObjectMeta{Name:web-allow-external2,GenerateName:,Namespace:default,SelfLink:/apis/cilium.io/v2/namespaces/default/ciliumnetworkpolicies/web-allow-external2,UID:601d7c35-9ccb-11e9-970f-080027efb264,ResourceVersion:106207,Generation:1,CreationTimestamp:2019-07-02 13:14:51 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{kubectl.kubernetes.io/last-applied-configuration: {\"apiVersion\":\"cilium.io/v2\",\"kind\":\"CiliumNetworkPolicy\",\"metadata\":{\"annotations\":{},\"name\":\"web-allow-external2\",\"namespace\":\"default\"},\"spec\":{\"egress\":[{}],\"endpointSelector\":{\"matchLabels\":{\"app\":\"nginx\"}},\"ingress\":[{\"fromCIDR\":[\"85.238.106.89/32\"],\"toPorts\":[{\"ports\":[{\"port\":\"6379\",\"protocol\":\"TCP\"}]}]},{\"fromCIDR\":[\"62.16.0.182/32\"],\"toPorts\":[{\"ports\":[{\"port\":\"80\",\"protocol\":\"TCP\"}]}]}]}}\n,},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,ManagedFields:[],}, Spec: {{\"matchLabels\":{\"any:app\":\"nginx\"}} [{[] [] [{[{6379 TCP}] <nil>}] [85.238.106.89/32] [] [] []} {[] [] [{[{80 TCP}] <nil>}] [62.16.0.182/32] [] [] []}] [{[] [] [] [] [] [] [] [] [] []}] [] }Status: {map[] map[]}" subsys=k8s
```

It does not matter whether the policy is submitted as a `CiliumNetworkPolicy` or a K8s `NetworkPolicy`, the rejection happens during rule validation here:

https://github.com/cilium/cilium/blob/81dfcb918a8d41a5ad6a3f46c770a666a3e475cc/pkg/policy/api/rule_validation.go#L92-L96

**Comment by mikefaraponov** [comment]
Oh, thx for your answer anyway, are you going to add it soon, or its more complicated feature than I thought?

**Comment by gandro** [comment]
I found the corresponding issue for the ToPorts/FromCIDR combination in ingress rules. The issue has a bite more details: https://github.com/cilium/cilium/issues/4129 (closing this one as a duplicate)

It looks like initial work on the feature has been merged, but it still needs proper tests before it can be safely enabled.

**Comment by joestringer** [comment]
Update: The above fix should be part of the upcoming v1.8.2 release or later releases.
